File: app/src/main/java/com/kinvey/sample/statusshare/fragments/RegisterFragment.java
/** 
 * Copyright (c) 2014 Kinvey Inc.
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 * 
 * http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations under
 * the License.
 * 
 */
package com.kinvey.sample.statusshare.fragments;

import android.os.Bundle;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.Gravity;
import android.view.KeyEvent;
import android.view.View;
import android.view.inputmethod.EditorInfo;
import android.widget.Button;
import android.widget.EditText;
import android.widget.TextView;
import android.widget.Toast;
import com.kinvey.android.model.User;
import com.kinvey.android.store.UserStore;
import com.kinvey.java.core.KinveyClientCallback;
import com.kinvey.sample.statusshare.MainActivity;
import com.kinvey.sample.statusshare.R;

/**
 * @author edwardf
 * @since 2.0
 */
public class RegisterFragment extends KinveyFragment implements View.OnClickListener {

    private EditText confirmPassword;
    private Button registerButton;
    private EditText userName;
    private EditText password;

    private TextView usernameLabel;
    private TextView passwordLabel;
    private TextView confirmPasswordLabel;

    private static final int MIN_USERNAME_LENGTH = 5;
    private static final int MIN_PASSWORD_LENGTH = 5;

    @Override
    public void onCreate(Bundle savedInstanceState){
        super.onCreate(savedInstanceState);
    }

    @Override
    public int getViewID() {
        return R.layout.fragment_register;
    }

    @Override
    public void bindViews(View v) {
        registerButton = (Button) v.findViewById(R.id.register_create_account);
        userName = (EditText) v.findViewById(R.id.register_username);
        password = (EditText) v.findViewById(R.id.register_password);
        confirmPassword = (EditText) v.findViewById(R.id.register_confirm_password);
        usernameLabel = (TextView) v.findViewById(R.id.register_username_label);
        passwordLabel = (TextView) v.findViewById(R.id.register_password_label);
        confirmPasswordLabel = (TextView) v.findViewById(R.id.register_confirm_label);

        usernameLabel.setTypeface(getRoboto());
        passwordLabel.setTypeface(getRoboto());
        confirmPasswordLabel.setTypeface(getRoboto());
        registerButton.setTypeface(getRoboto());
        userName.setTypeface(getRoboto());
        password.setTypeface(getRoboto());
        confirmPassword.setTypeface(getRoboto());

        this.addEditListeners();

    }

    private void addEditListeners() {

        registerButton.setOnClickListener(this);

        userName.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence charSequence, int i, int i1, int i2) {
            }

            @Override
            public void onTextChanged(CharSequence charSequence, int i, int i1, int i2) {
            }

            @Override
            public void afterTextChanged(Editable editable) {
                registerButton.setEnabled(validateInput());
            }
        });

        userName.setOnEditorActionListener(
                new EditText.OnEditorActionListener() {
                    @Override
                    public boolean onEditorAction(TextView v, int actionId, KeyEvent event) {
                        if ((actionId == EditorInfo.IME_ACTION_NEXT
                                || actionId == EditorInfo.IME_ACTION_DONE
                                || (event.getAction() == KeyEvent.ACTION_DOWN
                                && event.getKeyCode() == KeyEvent.KEYCODE_ENTER))
                                && userName.getText().length() < MIN_USERNAME_LENGTH
                                ) {

                            CharSequence text = "User name must contain at least " + MIN_USERNAME_LENGTH + " characters";
                            Toast.makeText(getActivity(), text, Toast.LENGTH_SHORT).show();
                            return true;
                        }
                        return false;
                    }
                });

        password.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence charSequence, int i, int i1, int i2) {
            }

            @Override
            public void onTextChanged(CharSequence charSequence, int i, int i1, int i2) {
            }

            @Override
            public void afterTextChanged(Editable editable) {
                registerButton.setEnabled(validateInput());
            }
        });

        password.setOnEditorActionListener(
                new EditText.OnEditorActionListener() {
                    @Override
                    public boolean onEditorAction(TextView v, int actionId, KeyEvent event) {
                        if ((actionId == EditorInfo.IME_ACTION_NEXT
                                || actionId == EditorInfo.IME_ACTION_DONE
                                || (event.getAction() == KeyEvent.ACTION_DOWN
                                && event.getKeyCode() == KeyEvent.KEYCODE_ENTER))
                                && password.getText().length() < MIN_USERNAME_LENGTH
                                ) {
                            CharSequence text = "Password must contain at least " + MIN_PASSWORD_LENGTH + " characters";
                            Toast.makeText(getActivity(), text, Toast.LENGTH_SHORT).show();
                            return true;
                        }
                        return false;
                    }
                });

        confirmPassword.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence charSequence, int i, int i1, int i2) {
            }

            @Override
            public void onTextChanged(CharSequence charSequence, int i, int i1, int i2) {
            }

            @Override
            public void afterTextChanged(Editable editable) {
                registerButton.setEnabled(validateInput());
            }
        });

        confirmPassword.setOnEditorActionListener(
                new EditText.OnEditorActionListener() {
                    @Override
                    public boolean onEditorAction(TextView v, int actionId, KeyEvent event) {
                        if ((actionId == EditorInfo.IME_ACTION_NEXT
                                || actionId == EditorInfo.IME_ACTION_DONE
                                || (event.getAction() == KeyEvent.ACTION_DOWN
                                && event.getKeyCode() == KeyEvent.KEYCODE_ENTER))
                                && (confirmPassword.getText().length() < MIN_USERNAME_LENGTH
                                || !password.getText().toString().equals(confirmPassword.getText().toString())
                        )) {
                            CharSequence text = "Repeat password must contain at least " + MIN_PASSWORD_LENGTH + " characters and equal password";
                            Toast.makeText(getActivity(), text, Toast.LENGTH_SHORT).show();
                            return true;
                        }
                        return false;
                    }
                });
    }

    private boolean validateInput() {
        return (userName.toString().length() >= MIN_USERNAME_LENGTH
                && password.getText().length() >= MIN_PASSWORD_LENGTH
                && confirmPassword.getText().length() >= MIN_PASSWORD_LENGTH
                && password.getText().toString().equals(confirmPassword.getText().toString()));
    }

    @Override
    public void onClick(View v) {
        if (v == registerButton){
            submit();
        }
    }

    public void submit() {
        UserStore.signUp(userName.getText().toString(), password.getText().toString(), getClient(), new KinveyClientCallback<User>() {
            public void onSuccess(User result) {
                if (getActivity() == null){
                    return;
                }
                CharSequence text = "Welcome " + result.get("username")+ ".";
                Toast.makeText(getActivity(), text, Toast.LENGTH_LONG).show();
                ((MainActivity) getActivity()).replaceFragment(new ShareListFragment(), false);
            }
            public void onFailure(Throwable error) {
                if (getActivity() == null){
                    return;
                }
                CharSequence text = "Username already exists.";
                Toast toast = Toast.makeText(getActivity(), text, Toast.LENGTH_LONG);
                toast.setGravity(Gravity.CENTER_VERTICAL, 0, 0);
                toast.show();
            }
        });
    }
}

File: app/src/main/java/com/kinvey/sample/statusshare/ui/fragments/RegisterFragment.kt
/**
 * Copyright (c) 2019 Kinvey Inc.
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations under
 * the License.
 *
 */
package com.kinvey.sample.statusshare.ui.fragments

import android.text.Editable
import android.text.TextWatcher
import android.view.KeyEvent
import android.view.View
import android.view.View.OnClickListener
import android.view.inputmethod.EditorInfo
import android.widget.TextView.OnEditorActionListener
import android.widget.Toast
import com.kinvey.android.Client
import com.kinvey.android.model.User
import com.kinvey.android.store.UserStore
import com.kinvey.java.core.KinveyClientCallback
import com.kinvey.sample.statusshare.R
import kotlinx.android.synthetic.main.fragment_register.*

/**
 * @author edwardf
 * @since 2.0
 */
class RegisterFragment : KinveyFragment(), OnClickListener {

    override val viewID: Int
        get() = R.layout.fragment_register

    override fun bindViews(v: View) {
        addEditListeners()
    }

    private val loginStr: String
        get() {
            return registerUsernameEdit?.text?.toString() ?: ""
        }

    private val passStr: String
        get() {
            return registerPasswordEdit?.text?.toString() ?: ""
        }

    private fun addEditListeners() {
        registerCreateAccountBtn?.setOnClickListener(this)
        registerUsernameEdit?.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(charSequence: CharSequence?, i: Int, i1: Int, i2: Int) {}
            override fun onTextChanged(charSequence: CharSequence?, i: Int, i1: Int, i2: Int) {}
            override fun afterTextChanged(editable: Editable?) {
                registerCreateAccountBtn?.isEnabled = validateInput()
            }
        })
        registerUsernameEdit?.setOnEditorActionListener(
            OnEditorActionListener { v, actionId, event ->
                if ((actionId == EditorInfo.IME_ACTION_NEXT || actionId == EditorInfo.IME_ACTION_DONE ||
                    (event.action == KeyEvent.ACTION_DOWN && event.keyCode == KeyEvent.KEYCODE_ENTER))
                  && registerUsernameEdit?.text?.length ?: 0 < MIN_USERNAME_LENGTH) {
                    val text: CharSequence = "User name must contain at least $MIN_USERNAME_LENGTH characters"
                    showToast(text)
                    return@OnEditorActionListener true
                }
                false
            })
        registerPasswordEdit?.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(charSequence: CharSequence?, i: Int, i1: Int, i2: Int) {}
            override fun onTextChanged(charSequence: CharSequence?, i: Int, i1: Int, i2: Int) {}
            override fun afterTextChanged(editable: Editable?) {
                registerCreateAccountBtn?.isEnabled = validateInput()
            }
        })
        registerPasswordEdit?.setOnEditorActionListener(
            OnEditorActionListener { v, actionId, event ->
                if ((actionId == EditorInfo.IME_ACTION_NEXT || actionId == EditorInfo.IME_ACTION_DONE ||
                    (event.action == KeyEvent.ACTION_DOWN && event.keyCode == KeyEvent.KEYCODE_ENTER))
                  && registerPasswordEdit?.text?.length ?: 0 < MIN_USERNAME_LENGTH) {
                    val text: CharSequence = "Password must contain at least $MIN_PASSWORD_LENGTH characters"
                    Toast.makeText(activity, text, Toast.LENGTH_SHORT).show()
                    return@OnEditorActionListener true
                }
                false
            })
        registerConfirmPasswordEdit?.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(charSequence: CharSequence?, i: Int, i1: Int, i2: Int) {}
            override fun onTextChanged(charSequence: CharSequence?, i: Int, i1: Int, i2: Int) {}
            override fun afterTextChanged(editable: Editable?) {
                registerCreateAccountBtn?.isEnabled = validateInput()
            }
        })
        registerConfirmPasswordEdit?.setOnEditorActionListener(
            OnEditorActionListener { v, actionId, event ->
                if ((actionId == EditorInfo.IME_ACTION_NEXT || actionId == EditorInfo.IME_ACTION_DONE ||
                    (event.action == KeyEvent.ACTION_DOWN && event.keyCode == KeyEvent.KEYCODE_ENTER))
                        && (registerConfirmPasswordEdit?.text?.length ?: 0 < MIN_USERNAME_LENGTH
                         || registerPasswordEdit?.text.toString() != registerConfirmPasswordEdit?.text.toString())) {
                    val text: CharSequence = "Repeat password must contain at least $MIN_PASSWORD_LENGTH characters and equal password"
                    Toast.makeText(activity, text, Toast.LENGTH_SHORT).show()
                    return@OnEditorActionListener true
                }
                false
            })
    }

    private fun validateInput(): Boolean {
        return registerUsernameEdit?.text?.length ?: 0 >= MIN_USERNAME_LENGTH
            && registerPasswordEdit?.text?.length ?: 0 >= MIN_PASSWORD_LENGTH
            && registerConfirmPasswordEdit?.text?.length ?: 0 >= MIN_PASSWORD_LENGTH
            && registerPasswordEdit?.text.toString() == registerConfirmPasswordEdit?.text.toString()
    }

    override fun onClick(v: View) {
        if (v === registerCreateAccountBtn) { submit() }
    }

    private fun submit() {
        UserStore.signUp(loginStr, passStr, client as Client<User>,
        object : KinveyClientCallback<User> {
            override fun onSuccess(result: User) {
                if (activity == null) { return }
                val text: CharSequence = "Welcome ${result["username"]}."
                showToast(text)
                replaceFragment(ShareListFragment(), false)
            }
            override fun onFailure(error: Throwable) {
                if (activity == null) { return }
                val text: CharSequence = "Username already exists."
                showToast(text, true)
            }
        })
    }

    companion object {
        private const val MIN_USERNAME_LENGTH = 5
        private const val MIN_PASSWORD_LENGTH = 5
    }
}